src/censored.jl

```diff
@@ -0,0 +1,519 @@
+end
+function censored(d::Censored, ::Nothing, u::Real)
+    return censored(d.uncensored, d.lower, d.upper === nothing ? u : min(u, d.upper))
+end
+function censored(d::Censored, l::Real, ::Nothing)
+    return censored(d.uncensored, d.lower === nothing ? l : max(l, d.lower), d.upper)
+end
+
+function params(d::Censored)
+    d0params = params(d.uncensored)
+    return (d0params..., d.lower, d.upper)
+end
+
+function partype(d::Censored{<:UnivariateDistribution,<:ValueSupport,T}) where {T}
+    return promote_type(partype(d.uncensored), T)
+end
+
+Base.eltype(::Type{<:Censored{D,S,T}}) where {D,S,T} = promote_type(T, eltype(D))
+
+#### Range and Support
+
+islowerbounded(d::LeftCensored) = islowerbounded(d.uncensored)
+islowerbounded(d::Censored) = islowerbounded(d.uncensored) || cdf(d.uncensored, d.lower) > 0
+
+isupperbounded(d::RightCensored) = isupperbounded(d.uncensored)
+isupperbounded(d::Censored) = isupperbounded(d.uncensored) || _ccdf_inclusive(d.uncensored, d.upper) > 0
+
+minimum(d::LeftCensored) = min(minimum(d.uncensored), d.upper)
+minimum(d::Censored) = max(minimum(d.uncensored), d.lower)
+
+maximum(d::RightCensored) = max(maximum(d.uncensored), d.lower)
+maximum(d::Censored) = min(maximum(d.uncensored), d.upper)
+
+function insupport(d::Censored, x::Real)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    return (
+        (_in_open_interval(x, lower, upper) && insupport(d0, x)) ||
+        (x == lower && cdf(d0, lower) > 0) ||
+        (x == upper && _ccdf_inclusive(d0, upper) > 0)
+    )
+end
+
+#### Show
+
+function show(io::IO, ::MIME"text/plain", d::Censored)
+    print(io, "Censored(")
+    d0 = d.uncensored
+    uml, namevals = _use_multline_show(d0)
+    uml ? show_multline(io, d0, namevals; newline=false) : show_oneline(io, d0, namevals)
+    if d.lower === nothing
+        print(io, "; upper=$(d.upper))")
+    elseif d.upper === nothing
+        print(io, "; lower=$(d.lower))")
+    else
+        print(io, "; lower=$(d.lower), upper=$(d.upper))")
+    end
+end
+
+_use_multline_show(d::Censored) = _use_multline_show(d.uncensored)
+
+
+#### Statistics
+
+quantile(d::Censored, p::Real) = _clamp(quantile(d.uncensored, p), d.lower, d.upper)
+
+median(d::Censored) = _clamp(median(d.uncensored), d.lower, d.upper)
+
+# the expectations use the following relation:
+# 𝔼_{x ~ d}[h(x)] = P_{x ~ d₀}(x < l) h(l) + P_{x ~ d₀}(x > u) h(u)
+#                 + P_{x ~ d₀}(l ≤ x ≤ u) 𝔼_{x ~ τ}[h(x)],
+# where d₀ is the uncensored distribution, d is d₀ censored to [l, u],
+# and τ is d₀ truncated to [l, u]
+
+function mean(d::LeftCensored)
+    upper = d.upper
+    log_prob_upper = logccdf(d.uncensored, upper)
+    prob_upper = exp(log_prob_upper)
+    μ = prob_upper * (iszero(prob_upper) ? oneunit(upper) : upper)
+    prob_trunc = exp(log1mexp(log_prob_upper))
+    # truncation contains ~ no probability
+    prob_trunc < eps(one(prob_trunc)) && return μ
+    return μ + oftype(μ, prob_trunc * mean(_to_truncated(d)))
```

Comment:
Can't/shouldn't we always use
```suggestion
    return μ + oftype(μ, prob_trunc * mean(_to_truncated(d)))
```
in case `mean(_to_truncated(d))` is very large and compensates for a small `prob_trunc`?

Reply:
Good point

Reply:
Not necessarily. `mean` is not defined for `Truncated`. In fact, this won't work for most distributions. But it's possible that a user censors a distribution in such a way that all probability mass falls outside of the censored interval. In that case the expectations are entirely determined by the atoms, so we don't even need `mean` to be implemented for the truncation. This is likely a rare edge case, but this check ensures that the expectations of `Censored` work for this edge case.

Reply:
We could still check for actual `iszero` without running risk of missing heavy but flat tails

Reply:
But what about the case I'm worried about? I don't mind if in some cases unnecessary computations are performed if it ensures that results are always correct.

Reply:
As noted in https://github.com/JuliaStats/Distributions.jl/pull/1470#discussion_r785777760, when we switched to computing log probabilities, the previous `iszero` checks stopped working. This check is not avoiding unnecessary computations. It's preventing an error being thrown in the edge case that 1) the truncation contains no probability mass and 2) `mean` is not implemented for the truncation. e.g.

```julia
julia> using Distributions
[ Info: Precompiling Distributions [31c24e10-a181-5473-b8eb-7969acd0382f]

julia> d0 = MixtureModel([Uniform(0, 1), Uniform(2, 3)]);

julia> d = censored(d0, 1.1, 1.9);

julia> mean(d)
```
With this check:
```julia
1.5
```

Without:
```julia
ERROR: MethodError: no method matching iterate(::Truncated{MixtureModel{Univariate, Continuous, Uniform{Float64}, Categorical{Float64, Vector{Float64}}}, Continuous, Float64})
Closest candidates are:
  iterate(::Union{LinRange, StepRangeLen}) at ~/packages/julias/julia-1.7/share/julia/base/range.jl:826
  iterate(::Union{LinRange, StepRangeLen}, ::Integer) at ~/packages/julias/julia-1.7/share/julia/base/range.jl:826
  iterate(::T) where T<:Union{Base.KeySet{<:Any, <:Dict}, Base.ValueIterator{<:Dict}} at ~/packages/julias/julia-1.7/share/julia/base/dict.jl:695
  ...
Stacktrace:
 [1] mean(f::typeof(identity), itr::Truncated{MixtureModel{Univariate, Continuous, Uniform{Float64}, Categorical{Float64, Vector{Float64}}}, Continuous, Float64})
   @ Statistics ~/packages/julias/julia-1.7/share/julia/stdlib/v1.7/Statistics/src/Statistics.jl:62
 [2] mean(itr::Truncated{MixtureModel{Univariate, Continuous, Uniform{Float64}, Categorical{Float64, Vector{Float64}}}, Continuous, Float64})
   @ Statistics ~/packages/julias/julia-1.7/share/julia/stdlib/v1.7/Statistics/src/Statistics.jl:44
 [3] mean(d::Distributions.Censored{MixtureModel{Univariate, Continuous, Uniform{Float64}, Categorical{Float64, Vector{Float64}}}, Continuous, Float64, Float64, Float64})
   @ Distributions ~/projects/GitHub/Distributions.jl/src/censored.jl:204
 [4] top-level scope
   @ REPL[7]:1
```

I welcome suggestions.

Reply:
You could still check `iszero(prob_trunc)`, can't you? This seems safe. In general though I tend to think it is fine to demand that `mean` is implemented for the untruncated distribution. More drastically, maybe one could even consider not supporting this edge case at all and throwing an error when calling `censored`. But I rather tend towards making `mean(d.untruncated)` a requirement.

Reply:
Isn't testing 
```
julia> logccdf(Uniform(), 1) == -Inf
true
```
as good?

Reply:
> You could still check `iszero(prob_trunc)`, couldn't you? This seems safe.

No, I found this fails even on trivial cases. But one can also cause a `DomainError` like this:
```julia
julia> log1mexp(logaddexp(log(0.2), log(0.8)))
ERROR: DomainError with -5.551115123125783e-17:
```
so perhaps this zero-truncation case is just too pathological and we should not try to special-case support it at all. I can drop the check. Might also allow some code simplifications.

Reply:
Yeah, the first check is meaningful if there is no mass beyond the truncation boundary. Failing if there is no mass inside the boundary *because* mean is not defined seems to be less pressing.

Reply:
(A bit unrelated: Are you aware of `logdiffcdf`? I guess it might not be too useful though since one again has to consider the special case of discrete distributions and add back the pmf value at the lower bound.)

Reply:
Thanks, I wasn't aware of it, but in this case it's not useful because we need the `logcdf` at both bounds, not just its difference.